server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -753,6 +766,11 @@ public String toString() {
         private final AckListener ackListener;
         private final ActionListener<Void> publishListener;
 
+        // We may not have accepted our own state before receiving a join from another node, causing its join to be rejected (we cannot
+        // safely accept a join whose last-accepted term/version is ahead of ours), so store them up and process them at the end.
+        // TODO this is unpleasant, is there a better way?
```

Comment:
Maybe it's not so bad. WDYT?

Reply:
as reconfiguration (which cares about the joins) can only happen in the next cluster state update (and is only triggered at the end of this publication), I think this is ok.